### ISSUE TITLE
actx.tag() comes too late to benefit eager mode execution

### ISSUE THREAD

**Issue by inducer** [inducer/meshmode#177]
For lazy mode, `actx.tag(MyTag(), 2*x+3)` works great. For eager mode, this doesn't help at all, as the result will already be computed (i.e. the kernel whose transformation would have benefited from the metadata was already executed). Is there an tagging interface that can work for both eager and lazy?

cc @alexfikl @thomasgibson @nchristensen @kaushikcfd 

x-ref: #176

**Comment by thomasgibson** [comment]
This is more or less kinda what I was trying (albeit rushed/incomplete) in https://github.com/inducer/meshmode/pull/170 --- Based on your comments there, I can imagine that one would tag information to a loopy program "as you go", which can then be transformed based on said information in [transform_loopy_kernel](https://github.com/inducer/meshmode/blob/main/meshmode/array_context.py#L683). This is assuming that, when doing any kind of eager evaluation, it always ends up going through [actx.call_loopy](https://github.com/inducer/meshmode/blob/main/meshmode/array_context.py#L656).

Obviously with eager this would be somewhat limited (since results are computed as you go), but _some_ level of code transformation can and should happen along the way.

With lazy, this information just gets propogated through until evaluation is needed, and so I imagine these transformations will occur at the pytato level.

**Comment by thomasgibson** [comment]
In short, I think tagging can still work for eager.

**Comment by inducer** [comment]
> if all expressions (whether eager or lazy) go through transform_loopy_kernel, then what can we do to potentially take advantage of this?

My thought was that lazy would bypass `transform_loopy_kernel` mostly. And much of eager (specifically the array arithmetic) also directly uses PyOpenCL array arithmetic and bypasses `transform_loopy_kernel`.

**Comment by inducer** [comment]
> In short, I think tagging can still work for eager.

I agree that it can, but what this issue is about is the method [`actx.tag`](https://documen.tician.de/meshmode/array.html#meshmode.array_context.ArrayContext.tag), whose job it is to apply a tag to an array result. The complaint is that it'll only work for lazy.
